### PR TITLE
lsm: support basic AppArmor stacking

### PR DESCRIPTION
**Pull request by brauner**
This is a revival of @tych0's patchset. I had to fix it up in quite a few places but with this I at least manage to restore the outer AppArmor namespace correctly. But I could use some good reviewers for this.
//Cc @tycho

Signed-off-by: Tycho Andersen <tycho.andersen@canonical.com>
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by adrianreber** [comment]
The test suite (zdtm) is not happy with your changes. The Python function you reference does not exist `NameError: name 'apparmor_entry' is not defined`

**Comment by brauner** [comment]
> The test suite (zdtm) is not happy with your changes. The Python function you reference does not exist `NameError: name 'apparmor_entry' is not defined`

Yeah, I quite expected that tbh.

**Comment by tych0** [comment]
Patches themselves look mostly reasonable, but I'm curious about this bit from the commit message:

> Support for apparmor namespaces and stacking is coming to Ubuntu kernels in
> 16.10, and should hopefully be upstreamed Soon (TM) :).

Has this landed upstream?

**Comment by brauner** [comment]
> Patches themselves look mostly reasonable, but I'm curious about this bit from the commit message:
> 
> > Support for apparmor namespaces and stacking is coming to Ubuntu kernels in
> > 16.10, and should hopefully be upstreamed Soon (TM) :).
> 
> Has this landed upstream?

Yeah, it has afaik. :)

**Comment by adrianreber** [comment]
The Alpine test failure is a known flake and there is already a fix for it open. I restarted it, sometimes it works.

**Comment by adrianreber** [comment]
Most CI runs are running on AppArmor enabled system. Can you add a zdtm test which verifies that AppArmor stacking checkpoint/restore works correctly? Especially to ensure we do not break it in the future.

**Comment by brauner** [comment]
> Most CI runs are running on AppArmor enabled system. Can you add a zdtm test which verifies that AppArmor stacking checkpoint/restore works correctly? Especially to ensure we do not break it in the future.

Yeah, I'll put that on my todo.

**Comment by rst0git** [comment]
> > Most CI runs are running on AppArmor enabled system. Can you add a zdtm test which verifies that AppArmor stacking checkpoint/restore works correctly? Especially to ensure we do not break it in the future.
> 
> Yeah, I'll put that on my todo.

There is test for AppArmor in Tycho's [apparmor-stacking-v4](https://github.com/tych0/criu/commits/apparmor-stacking-v4) branch: https://github.com/tych0/criu/commit/7ea7c0040e371381aa49984a4b6708e80ec29ac5

**Comment by brauner** [comment]
I'm happy to add the tests but I would like to get a rough "direction ok" or "sucks completely" from @avagin since he seems to be the only one who can merge right now? :)

**Comment by brauner** [comment]
Can I humbly suggest to add someone already well established in the #criu community as a maintainer with rights to merge things. This would probably help with the load of pull requests here. :)

**Comment by avagin** [comment]
@brauner Sorry for the delay. You are right, we have this problem. Usually, I review PR-s that have been approved by two maintainers with much higher priority. As for the rights to merge things, I will head up this question.

**Review by avagin — APPROVED** [review]

**Comment by avagin** [comment]
@adrianreber has got the right to merge PR-s. I quickly looked at these changes and they look good to me. Let's wait for Adrian's final review and probably this will be his first merged PR.

**Comment by adrianreber** [comment]
@brauner Thanks for recommending me.

The PR looks good. Please add the already mentioned test. Be aware that we also run tests on non AppArmor system. Maybe have a look at the SELinux tests which only run if SELinux is available.

**Comment by brauner** [comment]
> @brauner Thanks for recommending me.

No one to blame for that but yourself. :)

> 
> The PR looks good. Please add the already mentioned test. Be aware that we also run tests on non AppArmor system. Maybe have a look at the SELinux tests which only run if SELinux is available.

I'll add the tests and I'll fiddle the commit messages a bit and do another round of checks.

**Comment by adrianreber** [comment]
Looks good. There are a couple of PRs on the way to remove most of the remaining CI errors. I will ping you when they are merged and when you should rebase.

**Comment by adrianreber** [comment]
@avagin What do you day to the `run_command()` function? Does the criu codebase already have something like this?

File: criu/util.c
Comment (rst0git) [inline criu/util.c:1712]: There is a new function [read_all()](https://github.com/checkpoint-restore/criu/commit/178c90240f1bcd342c438d1d775aec00afd46ecd#diff-1120c47e8f3ca1a8b18111630a63190d0930ad39a2c63ff40b899add44872640R1592) in `until.c` that does something very similar.

File: criu/apparmor.c
Comment (rst0git) [inline criu/apparmor.c:508]: @0x7f454c46 opened this ticket some time ago: https://github.com/checkpoint-restore/criu/issues/767

File: criu/apparmor.c
Comment (rst0git) [inline criu/apparmor.c]: Could you please adjust the indentation on this line?

File: criu/apparmor.c
Comment (rst0git) [inline criu/apparmor.c:722]: It would be good if we could reduce the amount of BUG_ON() (https://github.com/checkpoint-restore/criu/issues/767).

File: criu/apparmor.c
Comment (rst0git) [inline criu/apparmor.c]: I'm not sure how unique apparmor namespaces are, but is it possible if the namespace already exists to be used by something else?

File: criu/util.c
Comment (brauner) [reply criu/util.c:1712]: This must be very new. I don't have that in my current dev branch. Good to know.

File: criu/apparmor.c
Comment (brauner) [reply criu/apparmor.c:508]: Ok, removing the BUG_ON.

File: criu/apparmor.c
Comment (brauner) [reply criu/apparmor.c:722]: Ok, removing the BUG_ON.

File: criu/apparmor.c
Comment (brauner) [reply criu/apparmor.c]: Fixed.

File: criu/apparmor.c
Comment (brauner) [reply criu/apparmor.c]: Yes, this is possible the assumption here is that we are almost always guaranteed to have multiple processes sharing the same namespaces so `write_aa_policy()` would always run into the `EEXIST` condition and it is perfectly fine for different application (independently snapshotted) to share the same AppArmor namespace.

File: criu/util.c
Comment (adrianreber) [inline criu/util.c:1664]: spaces

File: criu/util.c
Comment (adrianreber) [inline criu/util.c:1712]: hmm, sparc assembly code... not really needed for criu.

File: criu/util.c
Comment (adrianreber) [inline criu/util.c:1749]: interesting, in the other file you didn't align the struct initialization. but, that is mixed anyway all over the code base. so no need to change.

**Review by adrianreber — CHANGES_REQUESTED** [review]

File: criu/include/magic.h
Comment (Snorch) [inline criu/include/magic.h:102]: Not sure if the coordinates are right =) If I google 59°42'00.0"N 30°47'00.0"E I get to Nikolskoe not Sablino =) The closest "Sablino" I see  is at 59.64216480303743, 30.76562251056352 but yet it was renamed =) "Until 1922, the name of Ulyanovka was Sablino." https://en.wikipedia.org/wiki/Ulyanovka,_Leningrad_Oblast

**Review by Snorch — CHANGES_REQUESTED** [review]

File: criu/apparmor.c
Comment (Snorch) [inline criu/apparmor.c:240]: Recursion again, shouldn't we use ftw here?

File: criu/apparmor.c
Comment (Snorch) [inline criu/apparmor.c:555]: Recursion again... Can we use DFS over tree like we normally do. (see mnt_subtree_next() for an example)

**Review by Snorch — COMMENTED** [review]
> Switch to the run_command() helper introduced earlier which allows us to
> catch the output of the apparmor_parser command.

That `run_command()` is excess as cr_system is doing the same thing, see is_iptables_nft() for an example on how to read output from cr_system called command.

**Comment by brauner** [comment]
I'll try to get to rework this soonish.

File: criu/lsm.c
Comment (avagin) [reply criu/lsm.c:63]: it is allocated by fscanf, so free looks good here.

**Comment by avagin** [comment]
I fixed trivial comments and pushed this series to the criu-dev. Pls check that I didn't mess up anything.

File: criu/apparmor.c
Comment (avagin) [reply criu/apparmor.c:44]: this is a static function.

**Comment by brauner** [comment]
> I fixed trivial comments and pushed this series to the criu-dev. Pls check that I didn't mess up anything.

Thank you, @avagin. I appreciate you doing the fixups. Sorry for not getting around to it in time!

File: criu/apparmor.c
Comment (Snorch) [reply criu/apparmor.c:44]: grep may not know this =)

**Pull request closed without merging**
